### PR TITLE
Resolve some life cycle issues with `TabbedPage` when nested inside a `NavigationPage`

### LINKED ISSUE #7558: Setting SetIsSmoothScrollEnabled to false on a TabbedPage causes blank pages after navigating.
### Description

When setting SetIsSmoothScrollEnabled to false on a TabbedPage the page will work but after navigating between the tabs the content will disappear. 
Recording 2022-05-28 at 11 09 27

Example Project
https://github.com/joshm998/MAUITabbedPageIssueExample

### Steps to Reproduce

Created new .NET MAUI Project

Added in TabbedPage to start project, added some children to TabbedPage

set the following values 
`
On<Microsoft.Maui.Controls.PlatformConfiguration.Android>().SetIsSmoothScrollEnabled(false);
        On<Microsoft.Maui.Controls.PlatformConfiguration.Android>().SetIsSwipePagingEnabled(false);
`

Ran project and when navigating between pages after a while the content will disappear
image

### Version with bug

6.0 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android 11 and 12

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #9743: [Android] Content wrapped in NavigationPage objects in Tabbed Page disappears after tabs navigating
### Description

While navigating between 4 or more tabs that are wrapped in a 'NavigationPage' objects in 'TabbedPage', some tabs disappear. Instead of them blank Navigation Page is displayed and no errors are provided. OnAppearing() method is invoked but no contents are displayed.

On IOS 'TabbedPage' with 'NavigationPage' objects as children behaves as expected.

Reproduction project is [here](https://github.com/neordi/maui_tabbedpage_bug).

### Steps to Reproduce

1. Create a File > New .Net MAUI App
2. Replace 'ContentPage' with 'TabbedPage'
3. Add to 'TabbedPage' 4 children that are wrapped in a 'NavigationPage' objects
4. Launch Project on Android
5. Navigate from 1st tab to 4th and than to 1st again.

Content of 1st tab has disappeared.

### Version with bug

6.0.486 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android 11 and up

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #11259: Java.Lang.IllegalStateException: Can not perform this action after onSaveInstanceState
### Description

Exception on app resume on android, when pushing a modal page.
No exception on a similar scenario with xamarin forms.

This looks similar to https://github.com/xamarin/Xamarin.Forms/issues/1378.
Maybe you should simply change from `Commit()` to `CommitAllowingStateLoss()`, as done here: https://github.com/xamarin/Xamarin.Forms/pull/527.

### Steps to Reproduce

1. Create a maui application for android
2. Make the mainpage either a TabbedPage or a ContentPage
3. In application OnResume, push a modal page (call PushModalAsync)
4. Launch app, move app to background, then resume app. There is an exception and the modal page is not displayed.

### Link to public reproduction project repository

...

### Version with bug

6.0.486 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android 13

### Did you find any workaround?

_No response_

### Relevant log output

```shell
Java.Lang.IllegalStateException: Can not perform this action after onSaveInstanceState
   at Java.Interop.JniEnvironment.InstanceMethods.CallIntMethod(JniObjectReference instance, JniMethodInfo method, JniArgumentValue* args) in /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/obj/Release/net7.0/JniEnvironment.g.cs:line 11725
   at Java.Interop.JniPeerMembers.JniInstanceMethods.InvokeAbstractInt32Method(String encodedMember, IJavaPeerable self, JniArgumentValue* parameters) in /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/Java.Interop/JniPeerMembers.JniInstanceMethods_Invoke.cs:line 492
   at AndroidX.Fragment.App.BackStackRecord.Commit() in /Users/runner/work/1/s/generated/androidx.fragment.fragment/obj/Release/net6.0-android/generated/src/AndroidX.Fragment.App.BackStackRecord.cs:line 131
   at Microsoft.Maui.Controls.Handlers.TabbedPageManager.RemoveTabs() in D:\a\_work\1\s\src\Controls\src\Core\Platform\Android\TabbedPageManager.cs:line 186
   at Microsoft.Maui.Controls.Handlers.TabbedPageManager.OnTabbedPageDisappearing(Object sender, EventArgs e) in D:\a\_work\1\s\src\Controls\src\Core\Platform\Android\TabbedPageManager.cs:line 201
   at Microsoft.Maui.Controls.Page.SendDisappearing() in D:\a\_work\1\s\src\Controls\src\Core\Page.cs:line 520
   at Microsoft.Maui.Controls.Internals.NavigationModel.PushModal(Page page) in D:\a\_work\1\s\src\Controls\src\Core\NavigationModel.cs:line 204
   at Microsoft.Maui.Controls.Platform.ModalNavigationManager.PushModalAsync(Page modal, Boolean animated) in D:\a\_work\1\s\src\Controls\src\Core\Platform\ModalNavigationManager\ModalNavigationManager.Android.cs:line 117
   at Microsoft.Maui.Controls.Window.NavigationImpl.OnPushModal(Page modal, Boolean animated) in D:\a\_work\1\s\src\Controls\src\Core\HandlerImpl\Window\Window.Impl.cs:line 714

Java.Lang.IllegalStateException: Can not perform this action after onSaveInstanceState
   at Java.Interop.JniEnvironment.InstanceMethods.CallIntMethod(JniObjectReference instance, JniMethodInfo method, JniArgumentValue* args) in /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/obj/Release/net7.0/JniEnvironment.g.cs:line 11725
   at Java.Interop.JniPeerMembers.JniInstanceMethods.InvokeAbstractInt32Method(String encodedMember, IJavaPeerable self, JniArgumentValue* parameters) in /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/Java.Interop/JniPeerMembers.JniInstanceMethods_Invoke.cs:line 492
   at AndroidX.Fragment.App.BackStackRecord.Commit() in /Users/runner/work/1/s/generated/androidx.fragment.fragment/obj/Release/net6.0-android/generated/src/AndroidX.Fragment.App.BackStackRecord.cs:line 131
   at Microsoft.Maui.Controls.Platform.ModalNavigationManager.ModalContainer..ctor(IMauiContext windowMauiContext, Page modal, ViewGroup parentView) in D:\a\_work\1\s\src\Controls\src\Core\Platform\ModalNavigationManager\ModalNavigationManager.Android.cs:line 228
   at Microsoft.Maui.Controls.Platform.ModalNavigationManager.PresentModal(Page modal, Boolean animated) in D:\a\_work\1\s\src\Controls\src\Core\Platform\ModalNavigationManager\ModalNavigationManager.Android.cs:line 130
   at Microsoft.Maui.Controls.Platform.ModalNavigationManager.PushModalAsync(Page modal, Boolean animated) in D:\a\_work\1\s\src\Controls\src\Core\Platform\ModalNavigationManager\ModalNavigationManager.Android.cs:line 119
   at Microsoft.Maui.Controls.Window.NavigationImpl.OnPushModal(Page modal, Boolean animated) in
```

### LINKED ISSUE #11529: Fix MultiPageFragmentStateAdapter so it doesn't use `GetHashCode` for `GetItemId`
### Description

`GetHashCode` isn't guaranteed to be unique 

https://github.com/dotnet/maui/blob/35af6fb7dfec887f6ae72106916ef1e101d06371/src/Controls/src/Core/Platform/Android/MultiPageFragmentStateAdapter.cs#L32

We should also just reuse the `ShellFragmentStateAdapter` between both of these places because the `ShellFragmentStateAdapter` already uses a Unique `Id` count instead of `GetHashCode`

### PR DESCRIPTION
**Pull request by PureWeen**
### Description of Change

`NavigationViewFragment` needs to validate the `handler` on the incoming page before it reuses it. When an `Activity` was being destroyed and then re-created the same `handler` from the previous `Activity` was still being used. I basically just copied over the logic that was already part of `Shell` for handling this into non-Shell. 

During the process of writing tests for this PR I also fixed a few areas
- If the user inserts a `Tab` changing the position of the current selected item then we need to account for this
- If all the tabs get removed then the platform code should no longer crash
- modified the tests to test both top and bottom tab variations

### Workaround
https://github.com/PureWeen/ShanedlerSamples

#### Side note
I know there's an issue related to the crash this fixes but I can't find it :-/ 

### Issues Fixed

- Fixes #7558
- Fixes #9743
- Fixes #11259
- Fixes #11529

### CODE REVIEW

**Comment by rmarinho** [comment]
Is this one any relation to this ?

https://github.com/dotnet/maui/pull/9213

**Comment by NonameMissingNo** [comment]
@PureWeen The crash has been fixed by 7.0 ( #8009 )

**Comment by PureWeen** [comment]
> Is this one any relation to this ?
> 
> #9213

No, this has nothing to do with `Shell`.  This fixes an issue where `NavigationPage` isn't detecting correctly if it can reuse the `handler` on a pushed page. Currently it's always re-using it even if the `Activity` was destroyed. That being said :-) This was something I already fixed on `Shell` and I've basically just copied over that code to `NavigationPage`.  Everything comes back to Shell 😈

**Comment by PureWeen** [comment]
> @PureWeen The crash has been fixed by 7.0 ( #8009 )

This PR relates to `Android` not `Windows`

File: src/Controls/src/Core/Platform/Android/MultiPageFragmentStateAdapter.cs [hunk 1]
```diff
@@ -30,7 +34,125 @@ public override Fragment CreateFragment(int position)
...
+					if (sender is Page page)
+						page.HandlerChanged -= OnHandlerChanged;
+
+					return;
+				}
+
+				_platformView = _page.Handler?.PlatformView;
+
+				if (_platformView != null)
+					_page.HandlerChanged -= OnHandlerChanged;
```
Comment (mattleibow) [inline src/Controls/src/Core/Platform/Android/MultiPageFragmentStateAdapter.cs:153]: Should this be a `+=`? Re add the handler?

File: src/Controls/src/Core/Platform/Android/MultiPageFragmentStateAdapter.cs [hunk 2]
```diff
@@ -30,7 +34,125 @@ public override Fragment CreateFragment(int position)
...
+		void CheckItemKeys()
+		{
+			for (var i = 0; i < keys.Count; i++)
+			{
+				ItemKey item = keys[i];
+
+				if (!_page.Children.Contains(item.Page))
+				{
+					// Disconnect will remove the ItemKey from the keys list
+					item.Disconnect();
```
Comment (mattleibow) [inline src/Controls/src/Core/Platform/Android/MultiPageFragmentStateAdapter.cs:88]: Will this cause an issue with skipping?

For example, assume 3 items in an array:


```cs
var keys = new [] { 1, 2, 3 };
for (var i = 0; i < keys.Count; i++)
{
    WriteLine($"{i} = {keys[i]}");
    if (i == 2)
    {
        keys.RemoveAt(i);
        WriteLine($"remove, {keys.Count} left");
    }
    else
    {
        WriteLine($"no remove, {keys.Count} left");
    }
}
```

The output would be:

```
0 = 1
no remove, 3 left
1 = 2
remove, 2 left
```

The loop will exit as the item was removed, and now i++ == Count.

Might need to go backwards in this loop...

**Review by mattleibow — COMMENTED** [review]
19 of 27, this is big and exciting, but got some comments so far...

File: src/Controls/src/Core/Platform/Android/MultiPageFragmentStateAdapter.cs (hunk 2 above)
Comment (PureWeen) [reply src/Controls/src/Core/Platform/Android/MultiPageFragmentStateAdapter.cs:88]: 🤦fixed

File: src/Controls/src/Core/Platform/Android/MultiPageFragmentStateAdapter.cs (hunk 1 above)
Comment (PureWeen) [reply src/Controls/src/Core/Platform/Android/MultiPageFragmentStateAdapter.cs:160]: I added an additional comment. This event is just waiting for the first time when a handler is set. Once the platformview has been created for an ItemKey that's a permanent bond. If the handler changes then the key gets removed and a new key will get created. So, this event is purely here for the moment in time when we know a tab is connected to a page but the page hasn't been saturated.

File: src/Controls/src/Core/Platform/Android/MultiPageFragmentStateAdapter.cs [hunk 3]
```diff
@@ -30,7 +34,132 @@ public override Fragment CreateFragment(int position)
...
+					_page = null;
+				}
+
+				_platformView = null;
+			}
+
+			public void SetToStableView()
+			{
+				_stableView = true;
+			}
```
Comment (mattleibow) [inline src/Controls/src/Core/Platform/Android/MultiPageFragmentStateAdapter.cs:132]: Who calls this one?

**Review by mattleibow — COMMENTED** [review]
Looks good, seems to just be 1 question that I am seeing here, but otherwise looks OK.

File: src/Controls/src/Core/Platform/Android/MultiPageFragmentStateAdapter.cs (hunk 3 above)
Comment (PureWeen) [reply src/Controls/src/Core/Platform/Android/MultiPageFragmentStateAdapter.cs]: 🤦

Added a test and organized the code to be a bit more straight forward when a key becomes set in stone. 

https://github.com/dotnet/maui/compare/7a349efc710d9d336e98cd61e86dc7b5a731815a...cde3e739ce3ab4b7c4b24bbef8feffd1479feb96

File: src/Controls/src/Core/Platform/Android/MultiPageFragmentStateAdapter.cs [hunk 4]
```diff
@@ -1,14 +1,16 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
 using AndroidX.Fragment.App;
 using AndroidX.ViewPager2.Adapter;
 
 namespace Microsoft.Maui.Controls.Platform
 {
-	internal class MultiPageFragmentStateAdapter<[DynamicallyAccessedMembers(BindableProperty.DeclaringTypeMembers)] T> : FragmentStateAdapter where T : Page
+	internal class MultiPageFragmentStateAdapter<T> : FragmentStateAdapter where T : Page
```
Comment (mattleibow) [inline src/Controls/src/Core/Platform/Android/MultiPageFragmentStateAdapter.cs:9]: I am not sure we can remove that as it is used by the `MultiPage<T>` field - and that still uses the attribute. Maybe not needed, but we can confirm with linker things later.

**Review by mattleibow — APPROVED** [review]

**Comment by espenrl** [comment]
@mattleibow will this one be backported to .NET 7?

**Pull request merged**